routing/control_tower.go

```diff
@@ -75,7 +110,28 @@ func (p *controlTower) RegisterAttempt(paymentHash lntypes.Hash,
 func (p *controlTower) Success(paymentHash lntypes.Hash,
 	preimage lntypes.Preimage) error {
 
-	return p.db.Success(paymentHash, preimage)
+	// Lock hash to prevent the following race: payment status is updated to
+	// succeeded, new subscriber is registered and will immediately receive
+	// the succeeded status, this function sends out the succeeded
+	// notification again to the same subscriber.
+	p.hashMutex.Lock(string(paymentHash[:]))
```

Comment:
The reason the lock needs to be acquired before the db transaction is to prevent a new subscription happening after the db transaction, but before `notifyFinalEvent`, as is described in the comment above. Maybe I misunderstand you suggestion?

Reply:
From reading the comment that sounds alright though?

Payment status is updated to succeeded, new subscriber comes in and (acquire lock) retrieves succeed status and sends notification (release lock), this function (acquire lock)  finds no one to send notification to (release lock).

Or is there another order of actions that require us to hold it while we update the DB?

Reply:
Updated code to only use a global lock. It indeed seems to be sufficient.